app/master.js

```diff
@@ -0,0 +1,366 @@
+      .then(() => { logger.info('All workers killed.'); });
+  }
+
+  /**
+   * Reload single worker, resolves when worker is listening socket.
+   * @param {object} worker - worker instance defined by cluster module.
+   * @return {Promise} promise to kill a worker and start a new one.
+   */
+  static reloadWorker(worker) {
+    logger.info(`Reloading Worker#${worker.id}.`);
+    return Master.killWorker(worker)
+      .then(() => Master.forkWorker());
+  }
+
+  /** 
+   * Reload all fork workers and return Promise.
+   * @return {Promise} promise to reload all workers.
+   */
+  static reloadAllWorkers() {
+    logger.info('Restarting all workers...');
+    return Promise.each(_.values(cluster.workers), Master.reloadWorker)
+      .then(() => { logger.info('All workers restarted.'); });
+  }
+
+  /**
+   * Returns a file event watcher for a specific directory.
+   * @param {string} dir - relative directory from workspace that you want to listen.
+   * @return {Emitter} watcher for file events.
+   */
+  static createFileListener(dir = '.') {
+    logger.info(`Creating watcher for directory: ${dir}.`);
+
+    // @note Listeners are created recursively, check amount of created listeners before commiting
+    const options = {
+      ignored: /(^|[/\\])\.|node_modules/,
+      ignoreInitial: true
+    };
+
+    const watcher = fileWatcher.watch(dir, options);
+    watcher.on('ready', () => {
+      logger.debug(`File watcher ready, watchers included: ${Object.keys(watcher.getWatched()).length}`);
+    });
+
+    watcher.on('error', (error) => {
+      logger.error(`Chokidar reported an error: ${error.message}`);
+      logger.debug(error.stack);
+    });
+
+    return watcher;
+  }
+
+  /** 
+   * Start reacting to file changes in the app directory.
+   * Reloads all workers if changes are detected, which in turn
+   * allows for zero server downtime updates.
+   * @param {Emitter} watcher - emitter that reports file events.
+   */
+  static activateFileListener(watcher) {
+    logger.info('Activating file listener, reloading workers automatically if changes detected.');
+
+    const masterFiles = [
+      path.join('app', 'master.js'),
+      path.join('tools', 'logger'),
+      path.join('tools', 'eventBus'),
+      path.join('tools', 'cluster')
+    ];
+
+    const listenedEvents = [
+      'change',
+      'add',
+      'remove',
+      'unlinkDir',
+      'addDir'
+    ];
+
+    watcher.on('all', (event, filePath) => {
+      if (listenedEvents.indexOf(event) === -1) {
+        logger.debug(`File event detected ${event}: ${filePath}`);
+      } else if (masterFiles.indexOf(filePath) === -1) {
+        logger.info('File changed, need to reload workers...');
+        eventBus.emit('workerRestartNeeded', `file changed: ${filePath}`);
+      } else {
+        logger.info(`Internal files (${filePath}) changed, the whole server needs to reset!`);
+        eventBus.emit('systemRestartNeeded', `file changed: ${filePath}`);
```

Comment:
add listener (once..) for this in master and when this happens set flag which is visible in `getStats()` -api . Also would be good to have indicate for `getStats()` when workers restart is ongoing...